File: nth-digit.py
class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        number = 1
        num_len = 1
        count = 9

        while n > num_len * count:
            n -= num_len * count
            num_len += 1
            count *= 10
            number *= 10

        number += (n - 1) // num_len

        return int(str(number)[(n - 1) % num_len])

if __name__ == '__main__':
    s = Solution()
    print(s.findNthDigit(10))
    print(s.findNthDigit(21))
    print(s.findNthDigit(100))

File: nth-digit/nth-digit.js
/**
 * @param {number} n
 * @return {number}
 */
var findNthDigit = function(n) {
    let len = 1;
    let number = 1;
    let count = 9;

    while (n > count * len) {
        n -= count * len;

        len += 1;
        count *= 10;
        number *= 10;
    }

    number += Math.floor((n - 1) / len);

    return parseInt(number.toString()[(n - 1) % len]);
};

let main = () => {
  console.log(findNthDigit(10));
  console.log(findNthDigit(21));
  console.log(findNthDigit(100));
  console.log(findNthDigit(1000));
}

if (require.main === module) {
  main();
}
